lib/ChatBot.jsx

```diff
@@ -258,6 +258,19 @@ class ChatBot extends Component {
     return similarSteps.length > 0 ? similarSteps[similarSteps.length - 1] : null;
   };
 
+  evaluateExpression = evalExpression => {
+    const steps = [];
+
+    const { previousSteps, currentStep } = this.state;
+    previousSteps.forEach(step => {
+      steps[step.id] = step;
+    });
+    steps[currentStep.id] = currentStep;
```

Comment:
Why do we need to do this? Can't we just pass in the `steps` as an argument to this function (which, in turn, comes from `this.state`)?

Reply:
The `steps`  in `this.state` contains the overall flow of the chat, not how chat has progressed so far. Only `previousSteps` and `renderedSteps` contain how the chat has progressed so far -- with `renderedSteps` storing how the chat has been rendered or should be rendered and `previousSteps` storing what happened in `previousSteps`. Thus, `previousSteps` essentially stores the previous value of some variables like `{terms}`.

To answer your question as to why we need to do this: the `evalExpression` is going to contain statements like `steps["{variable}"].value = steps["{variable}"].value + 1`. To allow users that write `evalExpression`s to not bother with whether it is `previousSteps` or `steps` or `renderedSteps`, I just introduced a local variable `steps` that takes only the recent values from `previousSteps` and allows users of `evalExpression`s to update those variables without dealing with the nitty-gritty details of how this react-simple-chatbot  works.

However, I do agree with you when you ask the question: 'is it correct to mutate steps values like that?'. It is surely would be better to not mutate steps like that. There were some technical constraints that prevented me from doing so when I was working on the `json-step-values` branch; but I think it's possible to introduce new steps rather than mutate steps. I will see if it can be done without breaking the code.